src/coreclr/jit/lower.cpp

```diff
@@ -8179,8 +8185,354 @@ void Lowering::LowerIndir(GenTreeIndir* ind)
+// OptimizeForLdp: Record information about an indirection, and try to optimize
+// it by moving it to be adjacent with a previous indirection such that they
+// can be transformed into 'ldp'.
+//
+// Arguments:
+//    ind - Indirection to record and to try to move.
+//
+// Returns:
+//    True if the optimization was successful.
+//
+bool Lowering::OptimizeForLdp(GenTreeIndir* ind)
+{
+    if (!ind->TypeIs(TYP_INT, TYP_LONG, TYP_DOUBLE, TYP_SIMD16))
+    {
+        return false;
+    }
+
+    target_ssize_t offs   = 0;
+    GenTree*       addr   = ind->Addr();
+    FieldSeq*      fldSeq = nullptr;
+    comp->gtPeelOffsets(&addr, &offs, &fldSeq);
+
+    if (!addr->OperIsLocal())
+    {
+        return false;
+    }
+
+    bool result = false;
+
+    SavedIndir newInd;
+    newInd.Indir    = ind;
+    newInd.AddrBase = addr->AsLclVarCommon();
+    newInd.Offset   = offs;
+
+    int maxCount = min(m_blockIndirs.Height(), 8);
+    for (int i = 0; i < maxCount; i++)
+    {
+        SavedIndir&   prev      = m_blockIndirs.TopRef(i);
+        GenTreeIndir* prevIndir = prev.Indir;
+        if ((prevIndir == nullptr) || (prevIndir->TypeGet() != ind->TypeGet()))
+        {
+            continue;
+        }
+
+        if (GenTree::Compare(addr, prev.AddrBase))
+        {
+            JITDUMP("[%06u] and [%06u] are indirs off the same base with offsets +%03u and +%03u\n",
+                    Compiler::dspTreeID(ind), Compiler::dspTreeID(prevIndir), (unsigned)offs, (unsigned)prev.Offset);
+            if (abs(offs - prev.Offset) == genTypeSize(ind))
+            {
+                JITDUMP("  ..and they are amenable to ldp optimization\n");
+                if (TryMakeIndirsAdjacent(prevIndir, newInd.Indir))
+                {
+                    // Do not let the previous one participate in
+                    // another instance; that can cause us to e.g. convert
+                    // *(x+4), *(x+0), *(x+8), *(x+12) =>
+                    // *(x+4), *(x+8), *(x+0), *(x+12)
+                    prev.Indir = nullptr;
+                    result     = true;
+                    break;
+                }
+                break;
+            }
+            else
+            {
+                JITDUMP("  ..but at wrong offset\n");
+            }
+        }
+    }
+
+    m_blockIndirs.Push(newInd);
+    return result;
+}
+
+//------------------------------------------------------------------------
+// TryMakeIndirsAdjacent: Try to prove that it is legal to move an indirection
+// to be adjacent to a previous indirection. If successful, perform the move.
+//
+// Arguments:
+//    ind   - Indirection to record and to try to move.
+//    next  - [out] Next node to lower. Only written if the optimization was successful.
+//
+// Returns:
+//    True if the optimization was successful.
+//
+bool Lowering::TryMakeIndirsAdjacent(GenTreeIndir* prevIndir, GenTreeIndir* indir)
+{
+    if ((indir->gtFlags & GTF_IND_VOLATILE) != 0)
```

Comment:
Is this check needed even for the `prevIndir`?

Reply:
No, not for correctness, but I moved the check into `OptimizeForLdp` anyway which will have the effect of checking it on both indirs, since we won't be able to combine them if either is volatile.